library/swift/src/filters/AsyncRequestFilter.swift

```diff
@@ -0,0 +1,22 @@
+import Foundation
+
+/// RequestFilter supporting asynchronous resumption.
+public protocol AsyncRequestFilter: RequestFilter {
+  /// Invoked explicitly in response to an asynchronous resumeRequest() callback when filter
+  /// iteration has been stopped. The parameters passed to this invocation will be a snapshot
+  /// of any stream state that has not yet been forwarded along the filter chain.
+  ///
+  /// - return: The resumption status including any HTTP entities that will be forwarded.
+  func onResumeRequest(
+    headers: RequestHeaders?,
+    data: Data?,
+    trailers: RequestTrailers?,
+    endStream: Bool
+  ) -> FilterResumeStatus<RequestHeaders, RequestTrailers>
```

Comment:
Thinking about how users will consume this, I think they'll likely end up force-unwrapping the values they called `resume` with, which might be a little unwieldy. However, I suppose we can't really make these non-optional since `resumeRequest()` can be called after headers or data, in which case data will sometimes be nil. Probably nothing actionable here, just thinking out loud (though it might be good to list some examples in docs here)

Reply:
I've been thinking about this two, but short of having multiple different `onResume` invocations with overloaded signatures based on what's present, and their own specific return types, I can't see another way of supporting the different states. (Multiple `onResume`s would probably be annoying to implement also.)